### PR TITLE
Add [Available_filtering] pass

### PR DESCRIPTION
**Pull request by mshinwell**
This pull request adds a new pass that will only be used when generating enhanced debugging information.  It adjusts the contents of the register availability sets such as to discard registers that will not be used by the DWARF-generating code.  This could potentially be combined with one of the other passes in the future, but for the moment is left separate, not least to help with debugging.

### CODE REVIEW

File: asmcomp/debug/available_filtering.ml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    | Lend -> ()
+    | _ ->
+      insn.available_before
+        <- Reg_availability_set.canonicalise insn.available_before;
+      begin match insn.available_across with
+      | None -> ()
+      | Some available_across ->
+        insn.available_across
+          <- Some (Reg_availability_set.canonicalise available_across)
+      end;
```
Comment (damiendoligez) [inline asmcomp/debug/available_filtering.ml:31]: This pattern-matching could be replaced with
```
insn.available_across <- Misc.Option.map Reg_availability_set.canonicalise insn.available_across
```

**Review by damiendoligez — APPROVED** [review]
I've reviewed the last commit, and it's obviously safe to add to the compiler.

**Comment by mshinwell** [comment]
There's going to be another small addition to this one, so please don't merge yet.  Thanks for the review.

**Comment by mshinwell** [comment]
The means of dealing with this canonicalisation is being reworked; this pass will no longer be needed.

**Pull request closed without merging**